### ISSUE TITLE
Security: Load trusted mods first to provide them with a clean environment

### ISSUE THREAD

**Issue by appgurueu** [minetest/minetest#12182]
## Problem

The global environment a trusted mod is loaded in might already be compromised. Even the most basic builtin Lua functions might be overwritten to leak parts of the insecure environment. This makes properly securing trusted mods very cumbersome (if not impossible).

## Solutions

Always load the set of trusted mods (and their (possibly recursive) dependencies) first. In order for a mod to be loaded before a trusted mod it thus has to make its way to the trusted mods dependency chain.

## Alternatives

Provide secure mod environments, or directly load trusted mods in a shared insecure environment, with some way to still access the global environment currently used to expose mod APIs, all of which may not break compat.

**Comment by sfan5** [comment]
While I support doing this it is a small improvement at best. Securing Lua code running in a shared environment with untrusted code is close to impossible even if the trusted code gets to run first.

**Comment by appgurueu** [comment]
> While I support doing this it is a small improvement at best.

I think the effort to improvement ratio of this is quite decent, especially considering that the mod load order has to be reimplemented anyways.

**Comment by appgurueu** [comment]
> While I support doing this it is a small improvement at best. Securing Lua code running in a shared environment with untrusted code is close to impossible even if the trusted code gets to run first.

That's why I'm questioning the "shared environment": Trusted mods should get a different environment. It is currently trivially possible to hijack mod trust *even if mods keep it in a local variable inaccessible to other mods* in one of multiple ways:

* Overwrite the trusted mod files *at mod load time* (credits for this idea go to @KaiGerdMueller) by *hijacking into the loading trusted mod* either by tampering with the global table or by using `debug.sethook` (note: removing `debug.sethook` doesn't fix this *in the slightest*) - trust will be hijacked at next start;
* Override all of `loadfile`/`dofile` etc. to hijack the insecure environment as soon as it is passed from `init.lua` to any other loaded file;
* Hijack the insecure environment when it is temporarily stored in a global variable to pass it to other files;
* Hijack parts of the insecure environment when they are passed to overridden builtin functions

The latter three points *can* be mitigated by modders at the expense of code quality, but it's practically impossible to mitigate the first point. Removing the ability of mods to write anything to their mod folder at load time would break backwards compability.

In conclusion, mod trust is *de facto* "all or nothing". Mod security might as well be disabled altogether - instead of trusting specific mods - until trusted mods are loaded in an environment protected from untrusted mods.

**Comment by sfan5** [comment]
The first point is now solved by #14486.

Changing the loading order just came up in an internal discussion and opinions were generally in favor.
